buffer filter: Deprecate `max_request_time` timeout

auni53 opened the pull request:
*Description*: The new request_timeout in http connection manager covers the filter chain.
*Risk Level*: N/A
Fixes #4830

Reminder to whichever maintainer reviews this, please send notice to envoy-announce@.

zuercher commented:
I think the field should be marked as deprecated in the [proto](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/filter/http/buffer/v2/buffer.proto#L22) as well.

I'm not expert at the validation rules, but it seems like the required restriction needs to be relaxed? Otherwise you can't leave it out in favor of the HCM config.

mattklein123 commented on api/envoy/config/filter/http/buffer/v2/buffer.proto:
```diff
@@ -19,13 +19,9 @@ message Buffer {
-    (gogoproto.stdduration) = true
-  ];
+  // This is deprecated in favour of http conn manager request timeouts.
+  google.protobuf.Duration max_request_time = 2
```
Can you check the buffer filter code to make sure this won't crash now that it isn't required?

auni53 commented:
(looking into this)

auni53 commented:
OK, made some non-trivial changes:

1. Changed a test which used to assert a default-constructed buffer.proto would be invalid
2. Removed bufferfactory assertions about proto config. (It asserted the presence of max_request_bytes for some reason even though that's an optional field.)
3. Removed one more level of max_request_time validation at buffer_filter.cc

auni53 commented:
/retest

mattklein123 commented on source/extensions/filters/http/buffer/buffer_filter.cc:
```diff
@@ -23,8 +23,8 @@ BufferFilterSettings::BufferFilterSettings(
       max_request_bytes_(static_cast<uint64_t>(proto_config.max_request_bytes().value())),
-      max_request_time_(
-          std::chrono::seconds(PROTOBUF_GET_SECONDS_REQUIRED(proto_config, max_request_time))) {}
+      max_request_time_(std::chrono::seconds(
```
The timer here needs to become optional for the duration of the deprecation. Basically, either make this an absl::optional, or use a default of 0 to indicate the timer is disabled, then don't arm the timer in the rest of the code. If possible please also add a test for this case to make sure there is no timer.

auni53 replied:
This doesn't make an assertion that the timer value is set; the value should default to 0 when not set. I will look into testing to ascertain that.

mattklein123 replied:
My point is that you need to actually not enable the timer if it's not set. That is not done currently since the timer must be > 0 in the current code.

mattklein123 commented on source/extensions/filters/http/buffer/config.cc:
```diff
@@ -19,9 +19,6 @@ namespace BufferFilter {
 Http::FilterFactoryCb BufferFilterFactory::createFilterFactoryFromProtoTyped(
     const envoy::config::filter::http::buffer::v2::Buffer& proto_config,
     const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
-  ASSERT(proto_config.has_max_request_bytes());
```
Can you revert this and add required for the field in proto? I don't think it makes any sense for this to not be set? (Or perhaps `gt` requires it to be set? I'm not sure.)

zuercher commented on DEPRECATED.md:
```diff
@@ -16,6 +16,7 @@ A logged warning is expected for each deprecated item that is in deprecation win
 * Use of `runtime_key` in `RequestMirrorPolicy`, found in
   [route.proto](https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto)
   is deprecated. Set the `runtime_fraction` field instead.
+* Use of buffer filter `max_request_time` is deprecated. in favor of http connection manager of :ref:request timeouts <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.request_timeout>
```
Sorry, but I don't think the rst-style links work in this file -- needs to be markdown: `[linktext](url)`
